python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,327 @@
+from __future__ import annotations
+
+import ctypes
+import struct
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Annotated, ClassVar, Dict, List
+
+import numpy
+import tritonserver._api._dlpack as _dlpack
+from tritonserver import _c as triton_bindings
+
+DLPACK_DEVICE_TYPE_TO_TRITON_MEMORY_TYPE = defaultdict(
+    lambda: None,
+    {
+        _dlpack.DLDeviceType.kDLCUDA: triton_bindings.TRITONSERVER_MemoryType.GPU,
+        _dlpack.DLDeviceType.kDLCPU: triton_bindings.TRITONSERVER_MemoryType.CPU,
+    },
+)
+
+DLPACK_TO_TRITON_DTYPE = defaultdict(
+    lambda: triton_bindings.TRITONSERVER_DataType.INVALID,
+    {
+        (_dlpack.DLDataTypeCode.kDLBool, 1): triton_bindings.TRITONSERVER_DataType.BOOL,
+        (
+            _dlpack.DLDataTypeCode.kDLFloat,
+            16,
+        ): triton_bindings.TRITONSERVER_DataType.FP16,
+    },
+)
+
+
+NUMPY_TO_TRITON_DTYPE = defaultdict(
+    lambda: triton_bindings.TRITONSERVER_DataType.INVALID,
```

Comment:
question why we are using defaultdict instead of a regular dictionary?

Reply:
Should there be key error instead of returning "invalid" type value?